### PR TITLE
RE-1191 Control Log Storage

### PR DESCRIPTION
**Pull request by hughsaunders**
This PR adds two controls for log storage:

1. Ensure that no console log may exceed 200MB - the job will be killed if that limit is exceeded
1. Ensure that all jobs have a retention policy so logs are not stored indefinitely. A lint rule is added to enforce this. 

Issue: [RE-1191](https://rpc-openstack.atlassian.net/browse/RE-1191)

### CODE REVIEW

File: scripts/lint_naming_conventions.py [hunk 1]
```diff
@@ -89,6 +93,22 @@ def parse_file_name(in_dir, in_file):
     return 0
 
 
+# Ensure all jobs have a retention policy
+def check_retention(job, in_file):
```
Comment (git-harry) [inline scripts/lint_naming_conventions.py]: This function doesn't fail on `build-discarder: {}`, is that intentional? Based on [1] I'd have thought this would want to require either `days-to-keep` or `num-to-keep` is set. Is it worth putting limits on what these values can be?

[1] https://docs.openstack.org/infra/jenkins-job-builder/properties.html#properties.build-discarder

File: scripts/lint_naming_conventions.py [hunk 2]
```diff
@@ -89,6 +93,22 @@ def parse_file_name(in_dir, in_file):
     return 0
 
 
+# Ensure all jobs have a retention policy
```
Comment (git-harry) [inline scripts/lint_naming_conventions.py]: If you are going to add the functionality to this file, I think you should rename it and modify any references to it being specific to naming conventions because it and its use in lint.sh is a bit misleading.

File: pipeline_steps/common.groovy [hunk 3]
```diff
@@ -904,4 +906,20 @@ void setTriggerVars(){
   print ("Trigger: ${env.RE_JOB_TRIGGER} (${env.RE_JOB_TRIGGER_DETAIL})")
 }
 
+// add wrappers that should be used for all jobs.
+// max log size is in MB
+void globalWraps(Closure body){
+  // global timeout is long, so individual jobs can set shorter timeouts and
+  // and still have to cleanup, archive atefacts etc.
```
Comment (mattt416) [inline pipeline_steps/common.groovy]: Nit: s/atefacts/artifacts/

File: pipeline_steps/common.groovy [hunk 4]
```diff
@@ -904,4 +906,20 @@ void setTriggerVars(){
   print ("Trigger: ${env.RE_JOB_TRIGGER} (${env.RE_JOB_TRIGGER_DETAIL})")
 }
 
+// add wrappers that should be used for all jobs.
+// max log size is in MB
+void globalWraps(Closure body){
```
Comment (mattt416) [inline pipeline_steps/common.groovy]: Would it make sense to add another parameter for the timeout value, so that individual jobs can override on a case-by-case basis without having to add inner timeouts specific for the job?  If we do, I'd personally like to see the default (for when none specified) be far lower, like 1 hour.  This change would also allow us to remove some existing timeouts that are littered through jobs and common utils.

File: rpc_jobs/periodic_cleanup.yml [hunk 5]
```diff
@@ -40,64 +38,68 @@
...
-      for (int i=0; i<nodes.size(); i++){
-        nodeName = nodes[i]
-        parallel_steps['node_'+nodeName] = {
-          common.use_node(nodeName){
-            stage("Cleanup "+nodeName){
-              dir("rpc-gating") {
-                sh "scripts/workspace_cleanup.sh"
-                sh "scripts/tmp_cleanup.sh"
-                sh "scripts/docker_cleanup.sh"
+      library "rpc-gating@${RPC_GATING_BRANCH}"
```
Comment (git-harry) [inline rpc_jobs/periodic_cleanup.yml]: Why was this line moved? There is a pattern in place for having this as the first line, unless there is a good reason to move this I think it should be left where it is.

File: pipeline_steps/common.groovy [hunk 6]
```diff
@@ -729,7 +729,10 @@ void use_node(String label=null, body){
                          env.RPC_GATING_BRANCH)
       install_ansible()
       print "${env.NODE_NAME} preparation complete, now ready for use."
-      body()
+      // Ensure log size is not excessive
```
Comment (git-harry) [inline pipeline_steps/common.groovy]: Why is this needed here, why does `globalWraps` not suffice?

File: pipeline_steps/common.groovy [hunk 7]
```diff
@@ -904,4 +906,20 @@ void setTriggerVars(){
   print ("Trigger: ${env.RE_JOB_TRIGGER} (${env.RE_JOB_TRIGGER_DETAIL})")
 }
 
+// add wrappers that should be used for all jobs.
+// max log size is in MB
+void globalWraps(Closure body){
+  // global timeout is long, so individual jobs can set shorter timeouts and
+  // and still have to cleanup, archive atefacts etc.
+  timeout(time: 10, unit: 'HOURS'){
+    wrap([$class: 'LogfilesizecheckerWrapper', 'maxLogSize': 200, 'failBuild': true, 'setOwn': true]) {
```
Comment (git-harry) [inline pipeline_steps/common.groovy]: What does `'setOwn'` do?

File: scripts/lint_naming_conventions.py (hunk 1 above)
Comment (mattt416) [reply scripts/lint_naming_conventions.py]: Yep, I think we should check specific values ... with at minimum value > 0 being acceptable.

File: pipeline_steps/common.groovy (hunk 4 above)
Comment (git-harry) [reply pipeline_steps/common.groovy]: I think this is a good suggestion. I'd prefer to see it done as a separate patch. `globalWraps` allows for additional layers to be added painlessly. The timestamp and log size pieces don't duplicate anything existing. I'd suggest fixing this code without including any timeout changes then creating a separate pull request to address all the occurrences of timeout.

File: rpc_jobs/sync_credential_store.yml [hunk 8]
```diff
@@ -2,6 +2,9 @@
     name: 'Sync-Credential-Store'
     project-type: workflow
     concurrent: false
+    properties:
+      - build-discarder:
+          days-to-keep: 360
```
[the remark this answers is not in the document]
Comment (hughsaunders) [reply rpc_jobs/sync_credential_store.yml]: Because it hardly ever runs, doesn't log much, and the results could be useful when tracking down why a credentials does/doesn't exist.

File: pipeline_steps/common.groovy (hunk 6 above)
Comment (hughsaunders) [reply pipeline_steps/common.groovy]: Ooops, that was my first attempt before I added globalwraps. Will remove.

File: pipeline_steps/common.groovy (hunk 7 above)
Comment (hughsaunders) [reply pipeline_steps/common.groovy]: AFAICT it makes maxLogSize active, otherwise a value from Jenkins global config is used. I'd prefer to keep stuff out of Jenkins global config as its not in SCM.

File: rpc_jobs/periodic_cleanup.yml (hunk 5 above)
Comment (hughsaunders) [reply rpc_jobs/periodic_cleanup.yml]: This was to match the [phobos](https://github.com/rcbops/rpc-gating/blob/master/rpc_jobs/phobos.yml#L91) job where the functions are defined first, then the library statement is still the first line to get executed. 

I'm happy to put the library line back at the top, but should probably change that job as well.

File: pipeline_steps/common.groovy (hunk 6 above)
Comment (hughsaunders) [reply pipeline_steps/common.groovy]: Done

File: pipeline_steps/common.groovy (hunk 3 above)
Comment (hughsaunders) [reply pipeline_steps/common.groovy]: Done

File: rpc_jobs/periodic_cleanup.yml (hunk 5 above)
Comment (hughsaunders) [reply rpc_jobs/periodic_cleanup.yml]: Done

File: scripts/lint_naming_conventions.py (hunk 2 above)
Comment (hughsaunders) [reply scripts/lint_naming_conventions.py]: Done

**Comment by hughsaunders** [comment]
Comments have been addressed, now ready for review.

File: pipeline_steps/common.groovy (hunk 4 above)
Comment (hughsaunders) [reply pipeline_steps/common.groovy]: OK, Will do this as a separate patch.

Comment (hughsaunders) [reply pipeline_steps/common.groovy]: https://rpc-openstack.atlassian.net/browse/RE-1263

File: scripts/lint_jjb.py [hunk 9]
```diff
@@ -1,18 +1,21 @@
...
+#        https://github.com/rcbops/rpc-gating#naming-conventions
+#   * Retention Policy
 
 import argparse
 import os
 import re
 import sys
 import yaml
 
+import jmespath
```
Comment (mattt416) [inline scripts/lint_jjb.py]: Do we need to specify this in requirements/constraints?

File: lint.sh [hunk 10]
```diff
@@ -87,15 +88,20 @@ check_bash(){
...
   # not matching rpc-gating conventions
+  # Also runs other checks that require parsing JJB yaml including:
+  #   * Ensure retention policy is set for all jobs.
   dirs_to_lint="pipeline_steps,rpc_jobs,scripts"
   exclude_files="NonCPS.groovy"
-  python scripts/lint_naming_conventions.py \
+  python scripts/lint_jjb.py \
     --dirs ${dirs_to_lint} --exclude ${exclude_files} \
     && echo "Naming conventions: OK" \
     || { echo "Naming conventions: FAIL"; rc=1; }
```
Comment (mattt416) [inline lint.sh]: Probably worth updating these strings here.

File: scripts/lint_jjb.py (hunk 9 above)
Comment (hughsaunders) [reply scripts/lint_jjb.py]: Already in both, as used in other scripts.

File: lint.sh (hunk 10 above)
Comment (hughsaunders) [reply lint.sh]: Done

**Comment by hughsaunders** [comment]
Rebased and addressed comments, ready for re-review.

File: scripts/lint_jjb.py [hunk 11]
```diff
@@ -89,6 +96,67 @@ def parse_file_name(in_dir, in_file):
...
+def check_retention(job, in_file):
+    error_message = ("{f}/{j}: Valid build-discarder config not found."
+                     " Please ensure days-to-keep (0 < x <= 365) or"
+                     " num-to-keep (0 < x <= 100) is specified."
+                     " Properties: {p}\n"
+                     .format(j=job['name'], f=in_file,
+                             p=job.get('properties')))
+
+    error_counter = 0
+
```
Comment (git-harry) [inline scripts/lint_jjb.py]: A for-loop might be nicer here.

```python
checks = (
    ('days-to-keep', 0, 365),
    ('num-to-keep', 0, 100),
)

for (retention_type, min_retention, max_retention) in checks:
    try:
        check_retention_value(job, retention_type, min_retention, max_retention)
    except (TypeError, IndexError):
        return_value = 1
    except ValueError as e:
        return_value = 1
        print e
    else:
        return_value = 0
        break
else:
    sys.stderr.write(error_message)

return return_value
```

File: scripts/lint_jjb.py [hunk 12]
```diff
@@ -89,6 +96,67 @@ def parse_file_name(in_dir, in_file):
...
+        print e
+        error_counter += 1
+
+    if (error_counter > 1):
+        sys.stderr.write(error_message)
+        return 1
+
+    return 0
+
+
```
Comment (git-harry) [inline scripts/lint_jjb.py]: `out` defined on line 175 (I can't comment on that part of the file) should be updated so that it doesn't mislead on failure.

File: scripts/lint_jjb.py [hunk 13]
```diff
@@ -1,18 +1,21 @@
 #!/usr/bin/env python
 
-# Script used to review jenkins jobs and report job names
-# not following conventions
-# https://github.com/rcbops/rpc-gating#naming-conventions
+# Script used to review jenkins jobs for compliance with internal conventions:
+#   * Naming convetions:
```
Comment (git-harry) [inline scripts/lint_jjb.py]: conventions

File: lint.sh [hunk 14]
```diff
@@ -87,18 +88,23 @@ check_bash(){
   done < <(find ${fargs[@]} -iname \*.sh)
 }
 
-check_naming_standards() {
+# Check JJB for internal standards:
+#   * naming conventions
+#   * retention policy
```
Comment (git-harry) [inline lint.sh]: The above three lines look like they should be moved into the function. At the moment there are comments about the function either side the start of its definition.

File: lint.sh [hunk 15]
```diff
@@ -87,18 +88,23 @@ check_bash(){
...
+check_jjb_lint() {
   # Limits lint to only gating unique files
   # Excludes webhooktranslator as it has additional packaging
-  # and it's own tox
+  # and it's own tox.
   # Excludes NonCPS.groovy as this filename is required, but
   # not matching rpc-gating conventions
+  # Also runs other checks that require parsing JJB yaml including:
+  #   * Ensure retention policy is set for all jobs.
```
Comment (git-harry) [inline lint.sh]: These lines read like they were added before `check_naming_standards` was changed to `check_jjb_lint` and so should probably be removed.

File: lint.sh (hunk 14 above)
Comment (hughsaunders) [reply lint.sh]: Done

File: lint.sh [hunk 16]
```diff
@@ -87,18 +88,23 @@ check_bash(){
...
 
-check_naming_standards() {
+# Check JJB for internal standards:
+#   * naming conventions
+#   * retention policy
+check_jjb_lint() {
   # Limits lint to only gating unique files
   # Excludes webhooktranslator as it has additional packaging
-  # and it's own tox
+  # and it's own tox.
```
[the remark this answers is not in the document]
Comment (hughsaunders) [reply lint.sh]: Done

File: lint.sh (hunk 15 above)
Comment (hughsaunders) [reply lint.sh]: Done

File: pipeline_steps/common.groovy [hunk 17]
```diff
@@ -866,7 +862,7 @@ void setTriggerVars(){
 // max log size is in MB
 void globalWraps(Closure body){
   // global timeout is long, so individual jobs can set shorter timeouts and
-  // and still have to cleanup, archive atefacts etc.
+  // and still have to cleanup, archive artifacts etc.
```
[the remark this answers is not in the document]
Comment (hughsaunders) [reply pipeline_steps/common.groovy]: Done

File: scripts/lint_jjb.py (hunk 13 above)
Comment (hughsaunders) [reply scripts/lint_jjb.py]: Done

File: scripts/lint_jjb.py (hunk 11 above)
Comment (hughsaunders) [reply scripts/lint_jjb.py]: Much better thanks, shortcuts the second check if the first check is successful.

File: scripts/lint_jjb.py (hunk 12 above)
Comment (hughsaunders) [reply scripts/lint_jjb.py]: Done, also updated readme with retention policy requirement.

File: pipeline_steps/common.groovy [hunk 18]
```diff
@@ -685,10 +685,6 @@ void use_node(String label=null, body){
                          env.RPC_GATING_BRANCH)
       install_ansible()
       print "${env.NODE_NAME} preparation complete, now ready for use."
-      // Ensure log size is not excessive
-      wrap([$class: 'LogfilesizecheckerWrapper', 'maxLogSize': 200, 'failBuild': false, 'setOwn': true]) {
-        body()
-      }
```
[the remark this answers is not in the document]
Comment (hughsaunders) [reply pipeline_steps/common.groovy]: And it shouldn't have removed the body call ⚠️ 
Done.
